src/test/shell/integration/jvm_flags_escaping_test.sh

```diff
@@ -0,0 +1,273 @@
+  expect_log 'arg31=(C:\\L M )'
+  expect_log 'arg32=("C:\\l m ")'
+  expect_log 'arg33=(C:\\N O\\)'
+  expect_log 'arg34=("C:\\n o\\")'
+  expect_log 'arg35=(C:\\P Q\\ )'
+  expect_log 'arg36=("C:\\p q\\ ")'
+  expect_log 'arg37=(C:\\R\\S\\)'
+  expect_log 'arg38=(C:\\R x\\S\\)'
+  expect_log 'arg39=("C:\\r\\s\\")'
+  expect_log 'arg40=("C:\\r x\\s\\")'
+  expect_log 'arg41=(C:\\T U\\W\\)'
+  expect_log 'arg42=("C:\\t u\\w\\")'
+}
+
+function assert_jvm_flags() {
+  local -r enable_windows_escaping=$1
+
+  create_pkg
+
+  if $enable_windows_escaping; then
+    local -r flag="--incompatible_windows_escape_jvm_flags"
+    local -r expect_run_fails=false
+    local -r expect_tokenization_succeeds=false
+  else
+    local -r flag="--noincompatible_windows_escape_jvm_flags"
+    local -r expect_run_fails=$is_windows
+    local -r expect_tokenization_succeeds=$is_windows
+  fi
+
+  # Assert how jvm_flags are tokenized and passed to the Java program.
+  bazel build $flag foo:x --verbose_failures 2>$TEST_log \
+    || fail "expected success"
+  if $is_windows && $expect_run_fails; then
+    (RUNFILES_DIR= \
+     RUNFILES_MANIFEST_FILE= \
+     RUNFILES_MANIFEST_ONLY= \
+     bazel-bin/foo/x${EXE_EXT} &>$TEST_log ; ) \
+       && fail "expected failure" || true
+    expect_not_log "LAUNCHER ERROR"
+    expect_log "Could not find or load main class"
+  else
+    (RUNFILES_DIR= \
+     RUNFILES_MANIFEST_FILE= \
+     RUNFILES_MANIFEST_ONLY= \
+     bazel-bin/foo/x${EXE_EXT} &>$TEST_log ; ) || fail "expected success"
+    if $is_windows; then
+      expect_not_log "LAUNCHER ERROR"
+    fi
+    assert_output
+  fi
+
+  # Assert that a badly quoted jvm_flag cannot be Bash-tokenized.
+  if $expect_tokenization_succeeds; then
+    bazel build $flag foo:cannot_tokenize --verbose_failures 2>$TEST_log \
+      || fail "expected success"
+    (RUNFILES_DIR= \
+     RUNFILES_MANIFEST_FILE= \
+     RUNFILES_MANIFEST_ONLY= \
+     bazel-bin/foo/cannot_tokenize${EXE_EXT} &>$TEST_log ;) \
+       || fail "excepted success"
+  else
+    if $is_windows; then
+      # On Windows: badly quoted jvm_flags are caught at build time.
+      bazel build $flag foo:cannot_tokenize --verbose_failures 2>$TEST_log \
+        && fail "excepted failure" || true
+      expect_log "ERROR:.*in jvm_flags attribute of java_binary rule"
+    else
+      # On Linux/macOS/non-Windows: badly quoted jvm_flags are caught at run
+      # time.
+      bazel build $flag foo:cannot_tokenize --verbose_failures 2>$TEST_log \
+        || fail "expected success"
+      (RUNFILES_DIR= \
+       RUNFILES_MANIFEST_FILE= \
+       RUNFILES_MANIFEST_ONLY= \
+       bazel-bin/foo/cannot_tokenize${EXE_EXT} &>$TEST_log ;) \
+         && fail "excepted failure" || true
+      expect_log "syntax error near unexpected token"
+    fi
+  fi
+}
+
+function test_no_windows_jvm_flags_escaping() {
+  # $1=false: build with --noincompatible_windows_escape_jvm_flags
+  assert_jvm_flags false
```

Comment:
This assertion is too generic and makes the test a lot harder to read. Can you please it down?

Reply:
Good point! Done.